File: lib/weblog_migration/url_metadata.rb
# frozen_string_literal: true

require "cgi"
require "digest"
require "json"
require "pathname"
require "uri"
require "fileutils"

module WeblogMigration
  module UrlMetadata
    SAFE_ID = /\A[A-Za-z0-9][A-Za-z0-9._-]*\z/.freeze
    SAFE_FILENAME = SAFE_ID
    HTML_CONTENT_TYPES = %w[text/html application/xhtml+xml].freeze
    TEXT_LIMITS = { "title" => 300, "description" => 2_000 }.freeze

    module_function

    def fetch_url_metadata(manifest_path, output_path, assets_dir, report_path, timeout: 20.0, max_bytes: 5 * 1024 * 1024)
      raise ArgumentError, "timeout must be greater than zero" unless timeout.positive?
      raise ArgumentError, "max_bytes must be greater than zero" unless max_bytes.positive?

      entries = load_url_manifest(manifest_path)
      assets_dir = Pathname(assets_dir)
      FileUtils.mkdir_p(assets_dir)
      results = entries.map { |entry| fetch_url_entry(entry, assets_dir, timeout, max_bytes) }
      output_path = Pathname(output_path)
      FileUtils.mkdir_p(output_path.dirname)
      output_path.write(JSON.pretty_generate("assets" => results, "version" => 1) + "\n", encoding: "UTF-8")
      report = {
        "assets" => results.length,
        "fallback" => results.count { |result| result["status"] == "fallback" },
        "manifest_path" => manifest_path.to_s,
        "metadata_path" => output_path.to_s,
        "ready" => results.count { |result| result["status"] == "ready" },
        "results" => results,
        "run_at" => TimeFormat.iso8601(Time.now)
      }
      report_path = Pathname(report_path)
      FileUtils.mkdir_p(report_path.dirname)
      report_path.write(JSON.pretty_generate(report) + "\n", encoding: "UTF-8")
      report
    end

    def load_url_metadata(path)
      return {} if path.nil?

      path = Pathname(path)
      payload = JSON.parse(path.read(encoding: "UTF-8"))
      raw_assets = payload.is_a?(Hash) ? payload["assets"] : nil
      raise ArgumentError, "URL metadata is missing assets: #{path}" unless raw_assets.is_a?(Array)
      metadata = {}
      raw_assets.each do |raw_asset|
        raise ArgumentError, "URL metadata entry must be an object: #{path}" unless raw_asset.is_a?(Hash)
        asset_id = raw_asset["id"]
        url = raw_asset["url"]
        raise ArgumentError, "URL metadata has an unsafe asset ID: #{asset_id.inspect}" unless asset_id.is_a?(String) && asset_id.match?(SAFE_ID)
        raise ArgumentError, "URL metadata entry is missing URL: #{asset_id.inspect}" unless url.is_a?(String)
        raise ArgumentError, "URL metadata contains duplicate asset ID: #{asset_id.inspect}" if metadata.key?(asset_id)
        image = raw_asset["image"]
        if image
          local_path = image.is_a?(Hash) ? image["local_path"] : nil
          unless image.is_a?(Hash) && local_path.is_a?(String) && local_path.match?(SAFE_FILENAME)
            raise ArgumentError, "URL metadata image has an unsafe local path: #{asset_id.inspect}"
          end
        end
        metadata[asset_id] = raw_asset
      end
      metadata
    rescue Errno::ENOENT, Errno::EACCES, JSON::ParserError => error
      raise ArgumentError, "could not read URL metadata: #{path}: #{error.message}"
    end

    def load_url_manifest(path)
      valid_entries = Assets.load_manifest(path)
      payload = JSON.parse(Pathname(path).read(encoding: "UTF-8"))
      raw_assets = payload.is_a?(Hash) ? payload["assets"] : nil
      raise ArgumentError, "asset manifest is missing assets: #{path}" unless raw_assets.is_a?(Array)
      raw_by_id = raw_assets.to_h { |raw_asset| [raw_asset.fetch("id"), raw_asset] }
      valid_entries.filter_map do |entry|
        next unless entry["kind"] == "url"
        raw_asset = raw_by_id.fetch(entry.fetch("id"))
        source_post_ids = raw_asset.fetch("source_post_ids", [])
        unless source_post_ids.is_a?(Array) && source_post_ids.all? { |post_id| post_id.is_a?(String) }
          raise ArgumentError, "asset manifest source_post_ids must be strings: #{entry["id"].inspect}"
        end
        { "id" => entry.fetch("id"), "url" => entry.fetch("url"), "kind" => "url", "source_post_ids" => source_post_ids.dup }
      end
    rescue Errno::ENOENT, Errno::EACCES, JSON::ParserError => error
      raise ArgumentError, "could not read asset manifest: #{path}: #{error.message}"
    end

    def fetch_url_entry(entry, assets_dir, timeout, max_bytes)
      url = entry.fetch("url")
      result = {
        "description" => nil,
        "domain" => domain(url),
        "id" => entry.fetch("id"),
        "image" => nil,
        "image_url" => nil,
        "kind" => "url",
        "source_post_ids" => entry.fetch("source_post_ids").dup,
        "status" => "fallback",
        "title" => nil,
        "url" => url
      }

      begin
        response = HTTP.get(url, timeout:, max_bytes:)
        if response.status >= 400
          return failure_result(result, "HTTP #{response.status}: #{response.message}", http_status: response.status)
        end
        unless HTML_CONTENT_TYPES.include?(response.content_type)
          raise Assets::AssetFetchError, "expected text/html content, got #{response.content_type}"
        end
        parser = PageMetadataParser.new(response.body)
        result["http_status"] = response.status
        result["title"] = parser.og_title || parser.document_title
        result["description"] = parser.og_description
        unless parser.og_image.nil?
          image_url = URI.join(response.url, parser.og_image).to_s
          image_uri = URI.parse(image_url)
          unless %w[http https].include?(image_uri.scheme) && image_uri.host
            return failure_result(result, "og:image URL is not an HTTP(S) URL", keep_metadata: true)
          end
          result["image_url"] = image_url
          result["image"] = fetch_image(image_url, assets_dir, entry.fetch("id"), timeout, max_bytes)
          result["status"] = "ready"
        end
        result
      rescue HTTP::FetchError, Assets::AssetFetchError, URI::InvalidURIError, ArgumentError => error
        failure_result(result, error.message, keep_metadata: result["title"] || result["description"])
      end
    end

    def failure_result(result, error, http_status: nil, keep_metadata: false)
      result = result.dup
      result["status"] = "fallback"
      result["error"] = error
      result["http_status"] = http_status unless http_status.nil?
      unless keep_metadata
        result["title"] = nil
        result["description"] = nil
      end
      result
    end

    def fetch_image(url, assets_dir, asset_id, timeout, max_bytes)
      response = HTTP.get(url, timeout:, max_bytes:)
      if response.status >= 400
        raise Assets::AssetFetchError, "HTTP #{response.status}: #{response.message}"
      end
      unless response.content_type.start_with?("image/")
        raise Assets::AssetFetchError, "expected image content, got #{response.content_type}"
      end
      data = response.body
      local_path = "#{asset_id}#{Assets.file_suffix(url, response.content_type)}"
      assets_dir.join(local_path).binwrite(data)
      {
        "http_status" => response.status,
        "local_path" => local_path,
        "mime_type" => response.content_type,
        "sha256" => Digest::SHA256.hexdigest(data),
        "size" => data.bytesize,
        "url" => url
      }
    end

    def domain(url)
      URI.parse(url).host || URI.parse(url).to_s
    rescue URI::InvalidURIError
      url
    end

    def clean_text(value, limit)
      normalized = CGI.unescapeHTML(value.to_s).split.join(" ")
      normalized.empty? ? nil : normalized[0, limit]
    end

    class PageMetadataParser
      attr_reader :og_title, :og_description, :og_image

      def initialize(html)
        @og_title = nil
        @og_description = nil
        @og_image = nil
        @title_parts = []
        @in_title = false
        html.scan(/<meta\b[^>]*>/im) do |tag|
          attributes = parse_attributes(tag)
          property = (attributes["property"] || attributes["name"] || "").downcase
          content = attributes["content"]
          next if content.nil? || content.empty?
          case property
          when "og:title"
            @og_title ||= UrlMetadata.clean_text(content, TEXT_LIMITS.fetch("title"))
          when "og:description"
            @og_description ||= UrlMetadata.clean_text(content, TEXT_LIMITS.fetch("description"))
          when "og:image", "og:image:url"
            value = CGI.unescapeHTML(content).strip
            @og_image ||= value unless value.empty?
          end
        end
        html.scan(/<title\b[^>]*>(.*?)<\/title>/im) { |match| @title_parts << match.first }
      end

      def document_title
        UrlMetadata.clean_text(@title_parts.join, TEXT_LIMITS.fetch("title"))
      end

      private

      def parse_attributes(tag)
        tag.scan(/([:\w-]+)\s*=\s*(["'])(.*?)\2/m).to_h { |name, _quote, value| [name.downcase, value] }
      end
    end
  end
end

File: src/log_migration/url_metadata.py
import argparse
import hashlib
import json
import re
from datetime import datetime, timezone
from html.parser import HTMLParser
from pathlib import Path
from typing import Any, Sequence
from urllib.error import HTTPError, URLError
from urllib.parse import urljoin, urlsplit
from urllib.request import urlopen

from .assets import _AssetFetchError, _file_suffix, _load_manifest, _request_url


_SAFE_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")
_SAFE_FILENAME = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")
_HTML_CONTENT_TYPES = {"text/html", "application/xhtml+xml"}
_IMAGE_MIME_PREFIX = "image/"
_TEXT_LIMITS = {
    "title": 300,
    "description": 2_000,
}


def fetch_url_metadata(
    manifest_path: Path,
    output_path: Path,
    assets_dir: Path,
    report_path: Path,
    timeout: float = 20.0,
    max_bytes: int = 5 * 1024 * 1024,
) -> dict[str, Any]:
    """Fetch OGP metadata for URL assets without changing the source manifest."""

    if timeout <= 0:
        raise ValueError("timeout must be greater than zero")
    if max_bytes <= 0:
        raise ValueError("max_bytes must be greater than zero")

    entries = _load_url_manifest(Path(manifest_path))
    assets_dir = Path(assets_dir)
    assets_dir.mkdir(parents=True, exist_ok=True)
    results = [
        _fetch_url_entry(entry, assets_dir, timeout, max_bytes)
        for entry in entries
    ]
    payload = {
        "assets": results,
        "version": 1,
    }
    output_path = Path(output_path)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(
        json.dumps(payload, ensure_ascii=False, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )

    report = {
        "assets": len(results),
        "fallback": sum(result["status"] == "fallback" for result in results),
        "manifest_path": str(manifest_path),
        "metadata_path": str(output_path),
        "ready": sum(result["status"] == "ready" for result in results),
        "results": results,
        "run_at": datetime.now(timezone.utc).isoformat(),
    }
    report_path = Path(report_path)
    report_path.parent.mkdir(parents=True, exist_ok=True)
    report_path.write_text(
        json.dumps(report, ensure_ascii=False, indent=2, sort_keys=True) + "\n",
        encoding="utf-8",
    )
    return report


def load_url_metadata(path: Path | None) -> dict[str, dict[str, object]]:
    """Load and validate fetched URL metadata for the static renderer."""

    if path is None:
        return {}
    path = Path(path)
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise ValueError(f"could not read URL metadata: {path}") from error

    raw_assets = payload.get("assets") if isinstance(payload, dict) else None
    if not isinstance(raw_assets, list):
        raise ValueError(f"URL metadata is missing assets: {path}")

    metadata: dict[str, dict[str, object]] = {}
    for raw_asset in raw_assets:
        if not isinstance(raw_asset, dict):
            raise ValueError(f"URL metadata entry must be an object: {path}")
        asset_id = raw_asset.get("id")
        url = raw_asset.get("url")
        if not isinstance(asset_id, str) or not _SAFE_ID.fullmatch(asset_id):
            raise ValueError(f"URL metadata has an unsafe asset ID: {asset_id!r}")
        if not isinstance(url, str):
            raise ValueError(f"URL metadata entry is missing URL: {asset_id!r}")
        if asset_id in metadata:
            raise ValueError(f"URL metadata contains duplicate asset ID: {asset_id!r}")
        image = raw_asset.get("image")
        if image is not None:
            if not isinstance(image, dict):
                raise ValueError(f"URL metadata image must be an object: {asset_id!r}")
            local_path = image.get("local_path")
            if not isinstance(local_path, str) or not _SAFE_FILENAME.fullmatch(local_path):
                raise ValueError(
                    f"URL metadata image has an unsafe local path: {asset_id!r}"
                )
        metadata[asset_id] = dict(raw_asset)
    return metadata


def _load_url_manifest(path: Path) -> list[dict[str, object]]:
    entries = _load_manifest(path)
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise ValueError(f"could not read asset manifest: {path}") from error

    raw_assets = payload.get("assets") if isinstance(payload, dict) else None
    raw_by_id = {
        raw_asset.get("id"): raw_asset
        for raw_asset in raw_assets
        if isinstance(raw_asset, dict)
    }
    url_entries: list[dict[str, object]] = []
    for entry in entries:
        if entry["kind"] != "url":
            continue
        raw_asset = raw_by_id[entry["id"]]
        source_post_ids = raw_asset.get("source_post_ids", [])
        if not isinstance(source_post_ids, list) or not all(
            isinstance(post_id, str) for post_id in source_post_ids
        ):
            raise ValueError(
                f"asset manifest source_post_ids must be strings: {entry['id']!r}"
            )
        url_entries.append(
            {
                **entry,
                "source_post_ids": list(source_post_ids),
            }
        )
    return url_entries


def _fetch_url_entry(
    entry: dict[str, object],
    assets_dir: Path,
    timeout: float,
    max_bytes: int,
) -> dict[str, object]:
    asset_id = str(entry["id"])
    url = str(entry["url"])
    result: dict[str, object] = {
        "description": None,
        "domain": _domain(url),
        "id": asset_id,
        "image": None,
        "image_url": None,
        "kind": "url",
        "source_post_ids": list(entry["source_post_ids"]),
        "status": "fallback",
        "title": None,
        "url": url,
    }

    try:
        with urlopen(_request_url(url), timeout=timeout) as response:
            status, content_type, data = _read_response(response, max_bytes)
            if content_type not in _HTML_CONTENT_TYPES:
                raise _AssetFetchError(
                    f"expected text/html content, got {content_type}"
                )
            charset = response.headers.get_content_charset() or "utf-8"
            base_url = response.geturl() if hasattr(response, "geturl") else url
    except HTTPError as error:
        return _failure_result(result, f"HTTP {error.code}: {error.reason}", error.code)
    except (OSError, URLError, ValueError, _AssetFetchError) as error:
        return _failure_result(result, str(error))

    parser = _PageMetadataParser()
    parser.feed(data.decode(charset, errors="replace"))
    result["http_status"] = status
    result["title"] = parser.og_title or parser.document_title
    result["description"] = parser.og_description

    if parser.og_image:
        image_url = urljoin(base_url, parser.og_image)
        image_parts = urlsplit(image_url)
        if image_parts.scheme.lower() not in {"http", "https"} or not image_parts.netloc:
            return _failure_result(
                result,
                "og:image URL is not an HTTP(S) URL",
                keep_metadata=True,
            )
        result["image_url"] = image_url
        try:
            image = _fetch_image(image_url, assets_dir, asset_id, timeout, max_bytes)
        except HTTPError as error:
            return _failure_result(
                result,
                f"og:image HTTP {error.code}: {error.reason}",
                keep_metadata=True,
            )
        except (OSError, URLError, ValueError, _AssetFetchError) as error:
            return _failure_result(result, f"og:image: {error}", keep_metadata=True)
        result["image"] = image
        result["status"] = "ready"
    return result


def _failure_result(
    result: dict[str, object],
    error: str,
    http_status: int | None = None,
    *,
    keep_metadata: bool = False,
) -> dict[str, object]:
    result["status"] = "fallback"
    result["error"] = error
    if http_status is not None:
        result["http_status"] = http_status
    if not keep_metadata:
        result["title"] = None
        result["description"] = None
    return result


def _fetch_image(
    url: str,
    assets_dir: Path,
    asset_id: str,
    timeout: float,
    max_bytes: int,
) -> dict[str, object]:
    with urlopen(_request_url(url), timeout=timeout) as response:
        status, content_type, data = _read_response(response, max_bytes)
        if not content_type.startswith(_IMAGE_MIME_PREFIX):
            raise _AssetFetchError(
                f"expected image content, got {content_type}"
            )

    local_path = f"{asset_id}{_file_suffix(url, content_type)}"
    (assets_dir / local_path).write_bytes(data)
    return {
        "http_status": status,
        "local_path": local_path,
        "mime_type": content_type,
        "sha256": hashlib.sha256(data).hexdigest(),
        "size": len(data),
        "url": url,
    }


def _read_response(
    response: Any,
    max_bytes: int,
) -> tuple[int, str, bytes]:
    status = getattr(response, "status", None) or response.getcode()
    content_type = response.headers.get_content_type()
    content_length = response.headers.get("Content-Length")
    if content_length is not None and int(content_length) > max_bytes:
        raise _AssetFetchError(f"response exceeds {max_bytes} bytes")
    data = response.read(max_bytes + 1)
    if len(data) > max_bytes:
        raise _AssetFetchError(f"response exceeds {max_bytes} bytes")
    return status, content_type, data


def _domain(url: str) -> str:
    return urlsplit(url).hostname or urlsplit(url).netloc


def _clean_text(value: str, limit: int) -> str | None:
    value = " ".join(value.split())
    if not value:
        return None
    return value[:limit]


class _PageMetadataParser(HTMLParser):
    def __init__(self) -> None:
        super().__init__(convert_charrefs=True)
        self.og_title: str | None = None
        self.og_description: str | None = None
        self.og_image: str | None = None
        self._document_title_parts: list[str] = []
        self._in_title = False

    @property
    def document_title(self) -> str | None:
        return _clean_text("".join(self._document_title_parts), _TEXT_LIMITS["title"])

    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
        attributes = {
            name.lower(): value
            for name, value in attrs
            if value is not None
        }
        if tag.lower() == "meta":
            property_name = (
                attributes.get("property") or attributes.get("name") or ""
            ).lower()
            content = attributes.get("content")
            if not content:
                return
            if property_name == "og:title" and self.og_title is None:
                self.og_title = _clean_text(content, _TEXT_LIMITS["title"])
            elif property_name == "og:description" and self.og_description is None:
                self.og_description = _clean_text(
                    content,
                    _TEXT_LIMITS["description"],
                )
            elif property_name in {"og:image", "og:image:url"} and self.og_image is None:
                self.og_image = content.strip() or None
        elif tag.lower() == "title":
            self._in_title = True

    def handle_data(self, data: str) -> None:
        if self._in_title:
            self._document_title_parts.append(data)

    def handle_endtag(self, tag: str) -> None:
        if tag.lower() == "title":
            self._in_title = False


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Fetch OGP metadata for URL assets from a manifest"
    )
    parser.add_argument("--manifest", required=True, type=Path)
    parser.add_argument("--output", required=True, type=Path)
    parser.add_argument("--assets", required=True, type=Path)
    parser.add_argument("--report", required=True, type=Path)
    parser.add_argument("--timeout", type=float, default=20.0)
    parser.add_argument("--max-bytes", type=int, default=5 * 1024 * 1024)
    return parser


def main(argv: Sequence[str] | None = None) -> int:
    parser = build_parser()
    args = parser.parse_args(argv)
    try:
        fetch_url_metadata(
            args.manifest,
            args.output,
            args.assets,
            args.report,
            timeout=args.timeout,
            max_bytes=args.max_bytes,
        )
    except ValueError as error:
        parser.error(str(error))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
